buildtools/npm/manifest_test.go

```diff
@@ -1,65 +1,95 @@
 }
 
 func TestFromNodeModules(t *testing.T) {
-	t.Skip("not yet implemented")
-	manifests, err := npm.FromNodeModules("testdata/package.json")
+	testFromNodeModulesByFixture(t, "flattened_node_modules")
+	testFromNodeModulesByFixture(t, "nested_node_modules")
+}
+
+/*
+	├─┬ chai@4.1.2
+	│ ├── assertion-error@1.1.0
+	│ ├── check-error@1.0.2
+	│ ├─┬ deep-eql@3.0.1
+	│ │ └── type-detect@4.0.8
+	│ ├── get-func-name@2.0.0
+	│ ├── pathval@1.1.0
+	│ └── type-detect@4.0.8
+	└── type-detect@3.0.0
+*/
+
+func testFromNodeModulesByFixture(t *testing.T, fixture string) {
+	depGraph, err := npm.FromNodeModules(filepath.Join("fixtures", fixture))
 	assert.NoError(t, err)
 
-	/*
-		└─┬ chai@4.1.2
-			├── assertion-error@1.1.0
-			├── check-error@1.0.2
-			├─┬ deep-eql@3.0.1
-			│ └── type-detect@4.0.8 deduped
-			├── get-func-name@2.0.0
-			├── pathval@1.1.0
-			└── type-detect@4.0.8
-	*/
-
-	assert.NotEmpty(t, manifests)
-	assert.True(t, containsDep(manifests, "assertion-error", "1.1.0"), "Manifests does not include dep assertion-error")
-	assert.True(t, containsDep(manifests, "check-error", "1.0.2"), "Manifests does not include dep check-error")
-	assert.True(t, containsDep(manifests, "get-func-name", "2.0.0"), "Manifests does not include dep get-func-name")
-	assert.True(t, containsDep(manifests, "pathval", "1.1.0"), "Manifests does not include dep pathval")
-	assert.True(t, containsDep(manifests, "type-detect", "4.0.8"), "Manifests does not include dep type-detect")
-
-	// check transitive dep's existance
-	dep, err := selectDep(manifests, "deep-eql", "3.0.1")
-	assert.NoError(t, err, "Manifests does not include dep deep-eql")
-
-	assert.NotEmpty(t, dep.Dependencies)
-	assert.Len(t, dep.Dependencies, 1)
-	assert.Contains(t, dep.Dependencies, "type-detect")
-}
+	assert.Len(t, depGraph.Direct, 2)
+	chaiDirectDep := pkg.Import{
+		Target: "chai",
+		Resolved: pkg.ID{
+			Location: "",
+			Name:     "chai",
+			Revision: "4.1.2",
+			Type:     pkg.NodeJS,
+		},
+	}
 
-func selectDep(manifests []npm.Manifest, name string, version string) (npm.Manifest, error) {
-	for _, v := range manifests {
-		if v.Name == name && v.Version == version {
-			return v, nil
-		}
+	typeDetectDirectDep := pkg.Import{
+		Target: "type-detect",
+		Resolved: pkg.ID{
+			Location: "",
+			Name:     "type-detect",
+			Revision: "3.0.0",
+			Type:     pkg.NodeJS,
+		},
 	}
 
-	return npm.Manifest{}, errors.New("could not find manifest")
+	assert.Contains(t, depGraph.Direct, chaiDirectDep)
+	assert.Contains(t, depGraph.Direct, typeDetectDirectDep)
+
+	assert.Len(t, depGraph.Transitive, 8)
+
+	typeDetectTransitiveDepKey := pkg.ID{
+		Location: "",
+		Name:     "type-detect",
+		Revision: "4.0.8",
+		Type:     pkg.NodeJS,
+	}
+
+	// ensure project that is both a transitive and direct dep have correct versions based on where they live
+	assert.Contains(t, depGraph.Transitive, typeDetectTransitiveDepKey)
+
+	chaiProject := depGraph.Transitive[chaiDirectDep.Resolved]
```

Comment:
Can you also check the imports of `type-detect`? Checking the absence of bad data is just as important as the presence of good data.

At a higher level, it would be great if we had some facility for conveniently describing mock package graphs.